src/Features/CSharp/Portable/Completion/CompletionProviders/ImportCompletion/ImportCompletionProviderHelper.cs

```diff
@@ -3,18 +3,39 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.CSharp.Extensions;
 using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.CSharp.Completion.Providers
 {
     internal static class ImportCompletionProviderHelper
     {
-        public static ImmutableArray<string> GetImportedNamespaces(
-            SyntaxNode location,
-            SemanticModel semanticModel)
-            => semanticModel.GetUsingNamespacesInScope(location)
-                .SelectAsArray(namespaceSymbol => namespaceSymbol.ToDisplayString(SymbolDisplayFormats.NameFormat));
+        public static async Task<ImmutableArray<string>> GetImportedNamespacesAsync(SyntaxContext context, CancellationToken token)
+        {
+            // The location is the containing node of the LeftToken, or the compilation unit itself if LeftToken
+            // indicates the beginning of the document (i.e. no parent).
+            var location = context.LeftToken.Parent ?? context.SyntaxTree.GetRoot(token);
+            var usings = context.SemanticModel.GetUsingNamespacesInScope(location);
+
+            // Since we don't have a compiler API to easily get all global usings yet,
+            // hardcode the the name of SDK auto-generated global using file for now 
+            // as a temporary workaround.
+            var fileName = context.Document.Project.Name + ".GlobalUsings.g.cs";
+            var globalUsingDocument = context.Document.Project.Documents.FirstOrDefault(d => d.Name.Equals(fileName));
```

Comment:
I might be worried about nothing, and/or completely misunderstanding how Roslyn works, but it feels wasteful to scan for this document every time, when we know it will very rarely change, let alone change name. Can you cache the DocumentId, look up by that, and only scan if that lookup fails, or something? Is looking up by document id even faster? i have no idea 😁

Reply:
Yea, caching is a good idea.
@tmeschter Does the content of `.GlobalUsings.g.cs` ever change?, if not, then we don't even need to check the checksum for cached result.

Reply:
> Does the content of `.GlobalUsings.g.cs` ever change?

It's generated from MSBuild items (if I recall correctly) so the content could certainly change. For example, I can imagine having some UI in the property pages where the user can specify the set of global usings and the .g.cs file will update on the next design-time build.

I don't know if the name can change; I'm not familiar with how exactly the SDK generates the .g.cs file.

Reply:
Thanks @tmeschter. A quick search in SDK repo shows that the content of the file is based on `Using` items from msbuild, and the file name is hardcoded in SDK.

So based on this, I added simple cache to (hopefully) speed things up a little bit. Basically, we maintain a map between DocumentId and the checksum of document content plus global usings from it. Additionally there's a small optimization that when cache hits, we just return whatever in the cache and skip checksum comparison. A background task is triggered to do the check and update cache if needed.

Reply:
Shouldn't this also cache `globalUsingDocument` per project? I was thinking that was the main thing to avoid - spinning thought potentially thousands of Documents to find the one with the name we expect.

Reply:
Ah, OK. I somewhat misunderstood your suggestion. Sure, I can use ProjectId as the key to the map, and do the document look up as part of the slow path.

Reply:
Alright, fixed